SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c

```diff
@@ -164,21 +199,55 @@ HashCompleteAndExtend (
   HashCtx = (HASH_HANDLE *)HashHandle;
   ZeroMem (DigestList, sizeof (*DigestList));
 
-  for (Index = 0; Index < mHashInterfaceCount; Index++) {
-    HashMask = Tpm2GetHashMaskFromAlgo (&mHashInterface[Index].HashGuid);
-    if ((HashMask & PcdGet32 (PcdTpm2HashMask)) != 0) {
-      mHashInterface[Index].HashUpdate (HashCtx[Index], DataToHash, DataToHashLen);
-      mHashInterface[Index].HashFinal (HashCtx[Index], &Digest);
-      Tpm2SetHashToDigestList (DigestList, &Digest);
+  if (PcrIndex <= MAX_PCR_INDEX) {
+    for (Index = 0; Index < mHashInterfaceCount; Index++) {
+      HashMask = Tpm2GetHashMaskFromAlgo (&mHashInterface[Index].HashGuid);
+      if ((HashMask & PcdGet32 (PcdTpm2HashMask)) != 0) {
+        mHashInterface[Index].HashUpdate (HashCtx[Index], DataToHash, DataToHashLen);
+        mHashInterface[Index].HashFinal (HashCtx[Index], &Digest);
+        Tpm2SetHashToDigestList (DigestList, &Digest);
+      }
     }
-  }
 
   FreePool (HashCtx);
 
-  Status = Tpm2PcrExtend (
-             PcrIndex,
-             DigestList
-             );
+    Status = Tpm2PcrExtend (
+               PcrIndex,
+               DigestList
+               );
+  } else {
+    Status = Tpm2NvReadPublic (
+               PcrIndex,
+               &PublicInfo,
+               &PubName
+               );
+    if (EFI_ERROR(Status)) {
+      DEBUG ((DEBUG_ERROR, "TPM2 Nv 0x%x ReadPublic failed %r\n", PcrIndex, Status));
+      return Status;
+    }
+    DataSize = PublicInfo.nvPublic.dataSize;
+
+    for (Index = 0; Index < mHashInterfaceCount; Index++) {
+      HashMask = Tpm2GetHashMaskFromAlgo (&mHashInterface[Index].HashGuid);
+      AlgoId = Tpm2GetAlgoIdFromAlgo (&mHashInterface[Index].HashGuid);
+      if (((HashMask & PcdGet32 (PcdTpm2HashMask)) != 0)
+           && (AlgoId == PublicInfo.nvPublic.nameAlg)) {
+        mHashInterface[Index].HashUpdate (HashCtx[Index], DataToHash, DataToHashLen);
+        mHashInterface[Index].HashFinal (HashCtx[Index], &Digest);
+        Tpm2SetHashToDigestList (DigestList, &Digest);
+      }

```

Comment:
what if no match?

Reply:
add the below check if no match.
    if (DigestList->count == 0) {
      return EFI_NOT_FOUND;
    }